### PR TITLE
fix(roboticWarehouse): goal conditional

### LINKED ISSUE #190: Fix: robot warehouse rewarding non-goal states
The logic for checking whether a shelf is at a goal state upon delivery before providing a reward has a bug (i.e. shelves are being rewarded at none goal states). The suggested fix is to decompose the conditional (as was originally done).

### PR DESCRIPTION
**Pull request by arnupretorius**
Fix the goal conditional for reward in robotic warehouse.

Closes #190

### CODE REVIEW

File: jumanji/environments/routing/robot_warehouse/env.py [hunk 1]
```diff
@@ -491,7 +491,8 @@ def reward_and_update_request_queue_if_shelf_in_goal(
             return key, reward, request_queue, shelves
 
         # check if shelf is at goal position and in request queue
-        cond = (shelf_id != 0) & jnp.isin(shelf_id, request_queue + 1)
+        cond = jnp.logical_not(jnp.equal(shelf_id, 0))
+        cond = jnp.logical_and(cond, jnp.isin(shelf_id, request_queue + 1))
```
Comment (sash-a) [inline jumanji/environments/routing/robot_warehouse/env.py:495]: Could we change this to:

`shelf_at_goal = (~jnp.equal(shelf_id, 0)) & jnp.isin(shelf_id, request_queue + 1)`

Makes the `cond` variable more explicit and uses `~` and `&` instead of the jax functions

File: jumanji/environments/routing/robot_warehouse/env.py [hunk 2]
```diff
@@ -271,6 +265,12 @@ def update_state_scan(
             update_state_scan, (grid, agents, shelves, 0), actions
         )
 
+        # check for agent collisions
+        collisions = jax.vmap(functools.partial(utils.is_collision, grid))(
+            agents, self.agent_ids
+        )
```
Comment (sash-a) [inline jumanji/environments/routing/robot_warehouse/env.py:271]: Up to you, but I prefer

```suggestion
        collisions = jax.vmap(utils.is_collision, (None, 0, 0))(
            grid, agents, self.agent_ids
        )
```

File: jumanji/environments/routing/robot_warehouse/utils.py [hunk 3]
```diff
@@ -82,10 +82,11 @@ def get_valid_action(action_mask: chex.Array, action: chex.Array) -> chex.Array:
...
 
-def is_collision(grid: chex.Array, agent: Agent, action: int) -> chex.Array:
-    """Calculate whether an agent is about to collide with another
-    entity. If the agent instead collides with another agent, the
-    episode terminates (this behavior is specific to this JAX version).
+def is_collision(grid: chex.Array, agent: Agent, agent_id: int) -> chex.Array:
+    """Calculate whether an agent collides with another
+    agent. If there is a collision, theepisode terminates
+    (this behavior is specific to the JAX version of the
+    robotic warehouse (RWARE) environment).
```
Comment (sash-a) [inline jumanji/environments/routing/robot_warehouse/utils.py:89]:
```suggestion
    """Calculate whether an agent collides with another
    agent. If there is a collision, the episode terminates
    (this behavior is specific to the JAX version of the
    robotic warehouse (RWARE) environment).
```

**Review by sash-a — COMMENTED** [review]
Looks good, just some grammar and some really minor changes

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 1 above)
Comment (arnupretorius) [reply jumanji/environments/routing/robot_warehouse/env.py:495]: Thanks Sasha, when making these suggestions did you rerun the test? Since the issue fixed here crept in by accepting a similar suggestion, so I'm hesitant to do it again. Similar with the one suggested below. At this stage, since it is so minor, I'm inclined to keep what works. :)

Comment (clement-bonnet) [reply jumanji/environments/routing/robot_warehouse/env.py:495]: I lean towards @sash-a 's, yet both expressions are probably converted to the same XLA statement, so happy to keep as is.

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 2 above)
Comment (clement-bonnet) [reply jumanji/environments/routing/robot_warehouse/env.py:271]: I'm agnostic to this, up to you.

File: jumanji/environments/routing/robot_warehouse/env.py [hunk 4]
```diff
@@ -454,7 +454,7 @@ def _update_reward_and_request_queue(
         Returns:
             a random key, updated reward, request queue and shelves.
         """
-        x, y = goal
+        y, x = goal
```
Comment (clement-bonnet) [inline jumanji/environments/routing/robot_warehouse/env.py:457]: Was this an error as well? Why are coordinates inverted? In Jumanji, we usually follow the (x,y) convention instead of the video-game (y,x) one.

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 1 above)
Comment (clement-bonnet) [inline jumanji/environments/routing/robot_warehouse/env.py:495]: Why is this code being changed? The original looks better.

**Review by clement-bonnet — APPROVED** [review]
Looks good to me. Would it be possible to start a few steps of a2c training (with default params) and seed either 1, 2 or 3 to make sure the behavior is fixed?

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 1 above)
Comment (arnupretorius) [reply jumanji/environments/routing/robot_warehouse/env.py]: Reverted to Sasha's version

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 2 above)
Comment (arnupretorius) [reply jumanji/environments/routing/robot_warehouse/env.py:271]: Keeping as is.

File: jumanji/environments/routing/robot_warehouse/env.py (hunk 4 above)
Comment (arnupretorius) [reply jumanji/environments/routing/robot_warehouse/env.py:457]: Yes, this was a bug. The reversal is a quirk from the original environment in terms of how they set up the initial grid. See here on how they all invert when enumerating goal positions: https://github.com/semitable/robotic-warehouse/blob/7b83fe04ec5b57523bc4d6b6c1c179c48849a3f7/rware/warehouse.py#L807

**Comment by arnupretorius** [comment]
> Looks good to me. Would it be possible to start a few steps of a2c training (with default params) and seed either 1, 2 or 3 to make sure the behavior is fixed?

We ran some experiments to confirm, everything works as expected now. In essence, the problem remained the same but with inverted goal positions which caused a mismatch between the rendering. In the runs we did, the agents trained and the rendering looked good. I will also add an updated gif to this PR.

**Review by sash-a — APPROVED** [review]
LGTM :smile:

**Review by clement-bonnet — APPROVED** [review]
Thank you for the fixes! :rocket:

**Review by RuanJohn — APPROVED** [review]
Looks good! Thank you @arnupretorius :fire:
